contracts/src/interfaces/IStakeTable.sol

```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: Unlicensed
+
+pragma solidity ^0.8.0;
+
+import { BN254 } from "bn254/BN254.sol";
+import { EdOnBN254 } from "../libraries/EdOnBn254.sol";
+
+/// @title Interface for stake table that keep track of validators' external key and staked amount.
+/// @dev Stake delegation happens in a separate DelegationPool contract, and specific to
+/// instantiation thus not part of this interface
+interface IStakeTable {
+    /// @notice Supported stake type, either using native token or re-staked using ETH
+    enum StakeType {
+        Native,
+        Restake
+    }
+
+    /// @dev (sadly, Solidity doesn't support type alias on non-primitive types)
+    // We avoid declaring another struct even if the type info helps with readability,
+    // extra layer of struct introduces overhead and more gas cost.
+    //
+    /// @dev BLS verification key (used for Consensus voting in HotShot) is `BN254.G1Point`.
+    // `type BlsVerKey is BN254.G1Point;`
+    //
+    /// @dev Verification key of a Schnorr signature is just a `EdOnBN254.EdOnBN254Point`
+    // `type SchnorrVerKey is EdOnBN254.EdOnBN254Point;`
+
+    /// @notice Represents a HotShot validator node
+    /// In the dual-staking model, a HotShot validator could have multiple `Node` entries.
+    /// @param account The Ethereum account of the validator.
+    /// @param stakeType The type of token staked.
+    /// @param balance The amount of token staked.
+    /// @param registerEpoch The starting epoch for the validator.
+    /// @param exitEpoch The ending epoch for the validator.
+    /// @param schnorrVK The Schnorr verification key associated.
+    struct Node {
+        address account;
+        StakeType stakeType;
+        uint64 balance;
+        uint64 registerEpoch;
+        uint64 exitEpoch;
+        EdOnBN254.EdOnBN254Point schnorrVK;
+    }
+
+    // === Table State & Stats ===
+
+    /// @notice Get the commitment of the stake table used in current voting (i.e. snapshot at
+    /// the start of last epoch)
+    function votingStakeTableCommitment() external view returns (bytes32);
+    /// @notice Get the commitment of the stake table frozen for change (i.g. snapshot at the start
+    /// of the current epoch)
+    function frozenStakeTableCommitment() external view returns (bytes32);
+
+    /// @notice Get the total stakes of the registered keys in the latest stake table (Head).
+    /// @return The total stake for native token and restaked token respectively.
+    function totalStake() external view returns (uint256, uint256);
+    /// @notice Get the total number of stakers, uniquely identified by their `blsVK`
+    function totalKeys() external view returns (uint32);
+    /// @notice Get the total stake of the registered keys in the voting stake table
+    /// (LastEpochStart)
+    function totalVotingStake() external view returns (uint256);
```

Comment:
Similarly, do we need all of these onchain?

Reply:
when I add them, I try to think it from the perspective of "what normal users would expect a stake table API should offer" instead of "what's the minimum set of API our sequencer project need internally".

IMO, here's a conjured use case for the downstream app that might exist in our ecosystem:
a DAO that decide whether to stake more when its percentage drops below 10%. Then that contract would invoke these APIs to make its decision, right?

I think it'll be tough to require anyone who wants these basic stats to run the `espresso-sequencer` software and parse out the contract storage or history to read these values. WDYT?

Reply:
Sure, seems reasonable